File: Sources/OnScreen/Features/UfoAbduction.swift
import Foundation
import Yage

// MARK: - Schedule Event

extension ScreenEnvironment {
    func scheduleUfoAbduction() {
        scheduleRandomly(withinHours: 0..<5) { [weak self] in
            guard AppState.global.randomEvents else { return }
            guard let self, let victim = self.randomVictim() else { return }
            self.animateUfoAbduction(of: victim)
        }
    }
}

private extension ScreenEnvironment {
    func randomVictim() -> PetEntity? {
        children
            .compactMap { $0 as? PetEntity }
            .randomElement()
    }
    
    func animateUfoAbduction(of target: PetEntity) {
        let ufo = UfoEntity(in: self)
        children.append(ufo)
        ufo.abduct(target) {
            DispatchQueue.main.asyncAfter(deadline: .now() + 10) { [weak self] in
                self?.cleanUpAfterAbduction(of: target, by: ufo)
            }
        }
    }
    
    private func cleanUpAfterAbduction(of target: PetEntity, by ufo: PetEntity) {
        ufo.kill()
        target.kill()
        children.remove(ufo)
        children.remove(target)
        add(pet: target.species)
    }
}

// MARK: - Entity

private class UfoEntity: PetEntity {
    init(in world: World) {
        super.init(of: .ufo, in: world)
        frame.origin = world.bounds.topLeft
    }
    
    func abduct(_ target: Entity, onCompletion: @escaping () -> Void) {
        let abduction = UfoAbduction()
        install(abduction)
        abduction.abduct(target, onCompletion)
    }
}

// MARK: - Animation

private extension EntityAnimation {
    static let abduction = EntityAnimation(
        id: "abduction",
        size: CGSize(width: 1, height: 3),
        position: .entityTopLeft
    )
}

// MARK: - Abduction

private class UfoAbduction: Capability {
    private weak var target: Entity?
    private var subjectOriginalSize: CGSize!
    private var onCompletion: () -> Void = {}
    
    func abduct(_ target: Entity, _ onCompletion: @escaping () -> Void) {
        guard let subject else { return }
        subjectOriginalSize = subject.frame.size
        self.target = target
        self.onCompletion = onCompletion
        
        let seeker = Seeker()
        subject.install(seeker)
        let distance = CGSize(width: 0, height: -50)
        
        seeker.follow(target, to: .above, offset: distance) { [weak self] captureState in
            guard let self else {
                seeker.kill()
                onCompletion()
                return
            }
            guard case .captured = captureState else { return }
            seeker.isEnabled = false
            self.paralizeTarget()
            self.captureRayAnimation()
        }
    }
    
    private func paralizeTarget() {
        guard let target else { return }
        target.setGravity(enabled: false)
        target.direction = CGVector(dx: 0, dy: -1)
        target.speed = 1.2 * PetEntity.speedMultiplier(for: target.frame.size.width)
    }
    
    private func captureRayAnimation() {
        guard let subject, let target else { return }
        subject.direction = .zero
        subject.set(state: .action(action: .abduction, loops: 1))
        subject.capability(for: Seeker.self)?.kill()
        
        let shape = ShapeShifter()
        target.install(shape)
        shape.scaleLinearly(to: CGSize(width: 5, height: 5), duracy: 1.1)
        scheduleCompletion(in: 1.25)
    }
    
    private func scheduleCompletion(in delay: TimeInterval) {
        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            self?.animationCompleted()
        }
    }
    
    private func animationCompleted() {
        guard let target, let subject else { return }
        target.kill()
        subject.setGravity(enabled: false)
        subject.direction = CGVector(dx: 1, dy: -1)
        subject.set(state: .move)
        subject.movement?.isEnabled = true
        subject.frame.size = subjectOriginalSize
        onCompletion()
        kill()
    }
    
    override func kill(autoremove: Bool = true) {
        target = nil
        onCompletion = {}
        super.kill()
    }
}

private extension Species {
    static let ufo = Species(
        id: "ufo",
        capabilities: [
            "AnimatedSprite",
            "AnimationsProvider",
            "LinearMovement",
            "PetsSpritesProvider"
        ],
        dragPath: "front",
        movementPath: "front",
        speed: 2
    )
}

File: Sources/windows/onscreen/ufo_abduction.py
from datetime import datetime, timedelta
from onscreen.pet_entity import PetEntity
from yage.capabilities.gravity import Gravity
from yage.capabilities.seeker import Seeker, SeekerState, SeekerTargetPosition
from yage.capabilities.shape_shifter import ShapeShifter
from yage.models.animations import EntityAnimation, EntityAnimationPosition
from yage.models.entity import Entity
from yage.models.entity_state import EntityState
from yage.models.species import Species
from yage.utils.geometry import Point, Size, Vector
from yage.utils.logger import Logger

class UfoAbduction(PetEntity):
    def __init__(self, world):
        super().__init__(self._species(), world)
        Logger.log(self.id, "Created ufo...")
        self.frame.origin = Point(0, 0)
        self.is_ephemeral = True
        self._original_size = self.frame.size
        self._completion_date = None
        self._was_target_gravity_enabled = False
        self._kill_date = None

    @staticmethod
    def start(target: Entity):        
        ufo = UfoAbduction(target.world)
        target.world.children.append(ufo)
        ufo._abduct(target)

    def update(self, collisions, time):
        super().update(collisions, time)
        self._check_for_abduction_completion()
        self._check_for_event_completion()

    def _check_for_abduction_completion(self):
        if self._completion_date and self._completion_date < datetime.now():
            Logger.log(self.id, "Abduction completed", self.target.id)
            self._completion_date = None
            self._animation_completed()
    
    def _check_for_event_completion(self):
        if self._kill_date and self._kill_date < datetime.now():
            Logger.log(self.id, "All done!", self.target.id)
            self._kill_date = None
            self.kill()            

    def _abduct(self, target):
        Logger.log(self.id, "Started following", target.id)
        self.target = target
        self._original_target_size = target.frame.size
        seeker = self.capability(Seeker)
        seeker.follow(
            target,
            position=SeekerTargetPosition.ABOVE,
            offset=Point(0, -2.0*target.frame.height/3.0),
            auto_adjust_speed=False,
            on_update=self._on_seeker_update
        )

    def _on_seeker_update(self, status, distance):
        if status != SeekerState.CAPTURED: return
        Logger.log(self.id, "Captured", self.target.id)
        self.capability(Seeker).is_enabled = False
        self._paralize_target()
        self._capture_ray_animation()

    def _species(self) -> Species: 
        return Species(
            "ufo",
            [],
            [
                "AnimatedSprite",
                "AnimationsProvider",
                "LinearMovement",
                "PetsSpritesProvider",
                "Seeker"
            ],
            "front",
            10,
            "front",
            2,
            [],
            200
        )    

    def _paralize_target(self):
        Logger.log(self.id, "Paralized", self.target.id)
        if gravity := self.target.capability(Gravity): 
            self._was_target_gravity_enabled = gravity.is_enabled
            gravity.is_enabled = False
        self.target.direction = Vector(0, -1)
        self.target.reset_speed()
        self.target.speed = 1.2 * self.target.speed

    def _restore_target(self):
        Logger.log(self.id, "Restored", self.target.id)
        if gravity := self.target.capability(Gravity): 
            gravity.is_enabled = self._was_target_gravity_enabled
        self.target.direction = Vector(1, 0)
        self.target.frame.size = self._original_target_size
        self.target.place_in_random_position()

    def _capture_ray_animation(self):
        self.direction = Vector.zero()
        self.set_animation(self._abduction_animation(), 1)
        self.capability(Seeker).kill()
        shape = self.target.install(ShapeShifter)
        shape.scale_linearly(Size(1, 1), 1.1)
        self._schedule_completion(2.35)

    def _schedule_completion(self, delay):
        self._completion_date = datetime.now() + timedelta(seconds=delay)

    def _animation_completed(self):
        self.direction = Vector(1, -1)
        self.state = EntityState.MOVE
        self.movement.is_enabled = True
        self.frame.size = self._original_size
        self._kill_date = datetime.now() + timedelta(seconds=10)        

    def _abduction_animation(self):
        return EntityAnimation(
            'abduction', 
            Size(1, 3),
            EntityAnimationPosition.ENTITY_TOP_LEFT
        )

    def kill(self):
        self.world.children.remove(self)
        self._restore_target()
        self.target = None
        super().kill()